### PR TITLE
Initial

### PR DESCRIPTION
**Pull request by alexmojaki**
The point of this repo is to make the code [here](https://github.com/yola/userservice/blob/8ef8069d405b5902f4916d60fd82f5aa4722872f/userservice/management/commands/update_geoip_database.py) usable elsewhere, e.g. pixel.

Starting with the absolute basics. More PRs to follow.

### CODE REVIEW

File: maxmindupdater/__init__.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    except Exception, e:
+        # pyGeoIP could break in a variety of ways - we don't
+        # particularly care which ones.
+        sys.stderr.write('Retrieved invalid GeoIP database - '
+                         'check MaxMind account details: %s' % e)
+    else:
+        if not os.path.exists(os.path.dirname(curr_db)):
+            os.makedirs(os.path.dirname(curr_db))
+        move(new_db, curr_db)
+        os.rmdir(os.path.dirname(new_db))
```
Comment (stefanor) [inline maxmindupdater/__init__.py]: newline

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env python
+from setuptools import setup
+import maxmindupdater
+
+with open('README.rst') as readme_file:
+    readme = readme_file.read()
```
Comment (stefanor) [inline setup.py:7]: decode to unicode

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    license='MIT (Expat)',
+    url=maxmindupdater.__url__,
+    packages=['maxmindupdater'],
+    test_suite='nose.collector',
+    install_requires=[
+        'requests >= 2.0.0, < 3.0.0',
+    ],
+    classifiers=[
+        'License :: OSI Approved :: MIT License',
+        'Programming Language :: Python :: 2.7',
```
Comment (stefanor) [inline setup.py]: I'd try to aim for Python 3 support in new code, too

File: maxmindupdater/__init__.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            buf = f.read(block_size)
+    return hasher.hexdigest()
+
+
+def update_db(curr_db, license_key):
+    curr_db_path = os.path.dirname(curr_db)
+
+    r = requests.get(UPDATE_URL,
+                     params={'license_key': license_key,
+                             'edition_id': 'GeoIP2-Country',
```
Comment (stefanor) [inline maxmindupdater/__init__.py]: Presumably this should be a parameter. There are many maxmind DBs

File: maxmindupdater/__main__.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env python
+from argparse import ArgumentParser
+from maxmindupdater import update_db
+
+
+def main():
+    parser = ArgumentParser()
+    parser.add_argument('db_path')
+    parser.add_argument('license_key')
+    parser.add_argument('edition_id')
```
Comment (stefanor) [inline maxmindupdater/__main__.py]: These could be fleshed out a bit. Help strings and metavars, at least.

File: scripts/local_test.sh [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
```
Comment (stefanor) [inline scripts/local_test.sh]: wat

Seriously `/bin/bash` is a thing you can rely on. or `/bin/sh` even - this is all posix shell...

File: scripts/travis_test.sh [hunk 7]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+
+set -eux
+
+db_file=geoip.mmdb
+export PYTHONPATH=`pwd`
+set +x
+python maxmindupdater/__main__.py $db_file $LICENSE_KEY GeoIP2-Country
```
Comment (stefanor) [inline scripts/travis_test.sh]: `python -m maxmindupdater` ?

File: scripts/local_test.sh (hunk 6 above)
Comment (alexmojaki) [inline scripts/local_test.sh]: Haha I didn't notice. PyCharm inserts that automatically, not sure why.

File: setup.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env python
+from setuptools import setup
+import maxmindupdater
+
+with open('README.rst') as readme_file:
+    readme = readme_file.read().decode('utf8')
```
Comment (stefanor) [inline setup.py]: That probably isn't what you want to do on python3.

I'd suggest using `from codecs import open` and using its `encoding` parameter. That'll give you a consistent behaviour across python 2 and 3.

File: setup.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    entry_points={
+        'console_scripts': ['maxmind-updater=maxmindupdater.__main__:main'],
+    },
+    install_requires=[
+        'requests >= 2.0.0, < 3.0.0',
+    ],
+    classifiers=[
+        'License :: OSI Approved :: MIT License',
+        'Programming Language :: Python :: 2.7',
+        'Programming Language :: Python :: 3.4',
```
Comment (stefanor) [inline setup.py:37]: I think we're testing for more than those.

File: maxmindupdater/__main__.py (hunk 5 above)
Comment (alexmojaki) [inline maxmindupdater/__main__.py]: Why metavars? The names seem pretty clear. I don't even know what names I could give the metavars other than the existing names.

Comment (stefanor) [inline maxmindupdater/__main__.py]: Just so that they show that they take paramaters, in `--help`

Comment (alexmojaki) [inline maxmindupdater/__main__.py]:
```
$ python -m maxmindupdater --help
usage: __main__.py [-h] db_path license_key edition_id

positional arguments:
  db_path
  license_key
  edition_id

optional arguments:
  -h, --help   show this help message and exit
```

Comment (alexmojaki) [inline maxmindupdater/__main__.py]: Just did a test and metavars simply replace the names of positional arguments. So if I don't use a different name for the metavar they serve no purpose.

Comment (stefanor) [inline maxmindupdater/__main__.py]: Oh, right, they were args, not parameters. Yeah, never mind

File: .travis.yml [hunk 10]
```diff
@@ -0,0 +1,13 @@
+sudo: false
+language: python
+python:
+- '2.7'
+- '3.2'
```
Comment (adrianmoisey) [inline .travis.yml]: It looks like people are slowly dropping support for 3.2, which is EOL next month, so we could drop this?

File: .travis.yml [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+- '3.3'
+- '3.4'
+- '3.5'
+install: pip install -r requirements.txt
+script:
+- ./scripts/travis_test.sh
```
Comment (adrianmoisey) [inline .travis.yml]: I'm quite a fan of using `tox`, and then making travis-ci call `tox`. If you want to do this, checkout https://github.com/yola/demands as an example

File: scripts/local_test.sh [hunk 12]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
+# Used by Yola internally for testing (mainly to automatically obtain a license key)
+
+set -eux
+
+configurator --local userservice qa
+license_key=$(python -c "print __import__('json').load(open('configuration.json'))['userservice']['geoip']['license_key']")
+rm configuration.json
+
+./virtualenv/bin/python maxmindupdater/__main__.py geoip.mmdb $license_key GeoIP2-Country
```
Comment (stefanor) [inline scripts/local_test.sh]: `-m` here too?

File: maxmindupdater/__init__.py [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+def _hash_file(filename):
+    if not os.path.exists(filename):
+        return ''
+
+    block_size = 65536
+    hasher = hashlib.md5()
+
+    with open(filename, 'rb') as f:
+        buf = f.read(block_size)
+        while len(buf) > 0:
```
Comment (stefanor) [inline maxmindupdater/__init__.py]: There's a pythonic way to do this DRY:

``` python
for block in iter(lambda: f.read(block_size), ''):
```

Comment (alexmojaki) [inline maxmindupdater/__init__.py]: I think this feature of python is not well-known enough to be worth using. Many readers would be forced to look it up. I've used an alternative solution which is DRY and easily readable.

File: setup.py [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+        'requests >= 2.0.0, < 3.0.0',
+        'maxminddb >= 1.0.0, < 2.0.0'
+    ],
+    classifiers=[
+        'License :: OSI Approved :: MIT License',
+        'Programming Language :: Python :: 2.7',
+        'Programming Language :: Python :: 3.2',
+        'Programming Language :: Python :: 3.3',
+        'Programming Language :: Python :: 3.4',
+        'Programming Language :: Python :: 3.5',
```
Comment (stefanor) [inline setup.py:38]: IIRC you need to list `Programming Language :: Python :: 3` to be considered python3 compatible

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

File: scripts/local_test.sh (hunk 12 above)
Comment (alexmojaki) [inline scripts/local_test.sh]: I can't reproduce it but I sometimes got an error with `-m`.

File: setup.py [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    name='maxmindupdater',
+    version=metadata['version'],
+    description=metadata['doc'],
+    long_description=readme,
+    author='Yola',
+    author_email='engineers@yola.com',
+    license='MIT (Expat)',
+    url=metadata['doc'],
+    packages=['maxmindupdater'],
+    test_suite='nose.collector',
```
Comment (adrianmoisey) [inline setup.py]: I assume this is copypasta? Since this isn't using nose.

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

**Comment by adrianmoisey** [comment]
:+1:

**Pull request merged**